tests/e2e_tests/trap_pump_cal_e2e.py

```diff
@@ -0,0 +1,199 @@
+    ext_hdr['B_O'] = 0 # bias offset not simulated in the data, so set to 0;  -0.0394 DN from tvac_noisemap_original_data/results
+    ext_hdr['B_O_ERR'] = 0 # was not estimated with the II&T code
+
+    # Create a DetectorNoiseMaps object and save it
+    noise_maps = data.DetectorNoiseMaps(noise_map_dat, pri_hdr=pri_hdr, ext_hdr=ext_hdr,
+                                        input_dataset=mock_input_dataset, err=noise_map_noise,
+                                        dq=noise_map_dq, err_hdr=err_hdr)
+    noise_maps.save(filedir=trap_pump_outputdir, filename="mock_detnoisemaps.fits")
+    
+
+    # add calibration files to caldb
+    this_caldb = caldb.CalDB()
+    if e2e:
+        this_caldb.create_entry(nonlinear_cal)
+        this_caldb.create_entry(noise_maps)
+
+    ####### Run the walker on some test_data
+    if not e2e: # if you want to test older simulated data
+        walker.walk_corgidrp(trap_pump_data_filelist, "", trap_pump_outputdir, template="trap_pump_cal_small_size.json")
+    if e2e:
+        walker.walk_corgidrp(trap_pump_data_filelist, "", trap_pump_outputdir, template="trap_pump_cal.json")        
+
+
+    # clean up by removing entry
+    this_caldb.remove_entry(nonlinear_cal)
+    this_caldb.remove_entry(noise_maps)
+    # find cal file (naming convention for data.TrapCalibration class)
+    generated_trapcal_file = trap_cal_filename[:-5]+'_trapcal.fits'
+    generated_trapcal_file = os.path.join(trap_pump_outputdir, generated_trapcal_file) 
+    # Load
+    tpump_calibration = data.Image(generated_trapcal_file)
+    ##### Check against TVAC trap pump dictionary
+    
+    #Convert the output back to a dictionary for more testing. 
+    e2e_trap_dict = rebuild_dict(tpump_calibration.data)
+    e2e_trap_dict_keys = list(e2e_trap_dict.keys())
+
+    #Extract the extra info. 
+    unused_fit_data = tpump_calibration.ext_hdr['unfitdat']
+    unused_temp_fit_data = tpump_calibration.ext_hdr['untempfd']
+    two_or_less_count = tpump_calibration.ext_hdr['twoorles']
+    noncontinuous_count = tpump_calibration.ext_hdr['noncontc']
+    pre_sub_el_count = tpump_calibration.ext_hdr['prsbelct']
+
+    #####
+    # Run many of the tests from test_tfit_const_True_sub_noise_ill in ut_tpump_final.py
+    # these things are true of the TVAC result, so make sure they are also true of this output
+    assert(unused_fit_data > 0)
+    assert(unused_temp_fit_data == 0)
+    assert(two_or_less_count > 0)
+    assert(noncontinuous_count >= 0)
+    assert(pre_sub_el_count > 0)
+
+    for t in e2e_trap_dict_keys:
+        assert(t in TVAC_trap_dict)
+        
+    
+    #Note: removed several tests about sig_E and sig_cs, since we're not saving sig_E and sig_cs currently
+    for t in list(TVAC_trap_dict.keys()):
+        assert(t in e2e_trap_dict_keys)
+        assert((TVAC_trap_dict[t]['E'] == e2e_trap_dict[t]['E']) or 
+                ((TVAC_trap_dict[t]['E'] is None and np.isnan(e2e_trap_dict[t]['E']))))
```

Comment:
Hmm, it all works for me.  Let me look at my exact output values for this trap.

Reply:
So I'm still running into slightly different numbers. Which is mildly concerning that it might mean the code is not machine agnostic. However, the differences are all < 0.01%. 

I can get the tests to pass if I change the final chunk of code to be this: 

```    
for t in list(TVAC_trap_dict.keys()):
        assert(t in e2e_trap_dict_keys)
        assert(((TVAC_trap_dict[t]['E'] is None and np.isnan(e2e_trap_dict[t]['E']))) or 
               np.abs((TVAC_trap_dict[t]['E'] - e2e_trap_dict[t]['E'])/TVAC_trap_dict[t]['E']) < 1e-4)
        assert(((TVAC_trap_dict[t]['cs'] is None and np.isnan(e2e_trap_dict[t]['cs']))) or 
               np.abs((TVAC_trap_dict[t]['cs']-e2e_trap_dict[t]['cs'])/TVAC_trap_dict[t]['cs']) < 1e-4)
        assert(((TVAC_trap_dict[t]['tau at input T'] is None and np.isnan(e2e_trap_dict[t]['tau at input T']))) or 
               np.abs((TVAC_trap_dict[t]['tau at input T']-e2e_trap_dict[t]['tau at input T'])/TVAC_trap_dict[t]['tau at input T']))```

What I'm doing here is looking at the relative error and checking to see that the difference is <1e-4. 

Thoughts?

Reply:
Hey Max, are you on a Mac?  Python 3.12?  

At the bottom of this post are the results I got when I printed out the side-by-side comparison of E, cs, and tau at input T for the TVAC results (loaded in from Box) and the DRP results acting on the same data, and I get that they are exactly the same.  The numbers on the left are from the TVAC dictionary and should be exactly what you have when you load in that dictionary from Box.  Is that the case?

Actually, let me just print out the dictionary here so you can confirm you have the same thing.  If you open a terminal and do the following lines:

trap_pump_datadir = os.path.join(tvacdata_path, 'TV-20_EXCAM_noise_characterization', 'simulated_e2e_trap_pumped_frames')
sim_traps = os.path.join(tvacdata_path, 'TV-20_EXCAM_noise_characterization', "results", "tpump_e2e_results.npy")
# this is a .npy file; read it in as a dictionary
td = np.load(sim_traps, allow_pickle=True)
TVAC_trap_dict = dict(td[()])
print(TVAC_trap_dict)

you should have exactly this:
{((60, 80), 'LHSel2', 0): {'T': [210.0], 'tau': [0.00026383534656907034], 'sigma_tau': [9.655400171470195e-06], 'cap': [[0.9917007580757848, 0.02270235807196836, 6548.483005924797, 0.9857691733456044, 0.006733364398010327, 12190.264503732513, 0.9908950836190112, 0.013444029132542543, 22492.99179222869]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((10, 10), 'LHSel4', 0): {'T': [200.0, 210.0], 'tau': [1.1308675839120732e-05, 4.699204991280072e-06], 'sigma_tau': [5.388707673020261e-07, 2.525595275631441e-07], 'cap': [[1.0351176248850447, 0.0149669418601401, 13456.245447003697, 1.0248260567510865, 0.04712515874065606, 7083.715422301953, 1.010239694237208, 0.008410170484044403, 24067.765267526112], [0.9878758389865424, 0.027942368731741832, 12937.949991092491, 0.9740085428708097, 0.02896874657733853, 7355.563058799376, 0.9831612702812143, 0.005418341165293744, 24281.651574567823]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((26, 28), 'CENel1', 0): {'T': [200.0, 210.0, 220.0], 'tau': [0.00070055187417598, 0.00025516106214798024, 0.0001056539545833924], 'sigma_tau': [3.123879320692088e-05, 1.7559726893133466e-05, 4.867897647249574e-06], 'cap': [[1.0117639102281824, 0.03836781750864818, 6739.294152824368, 1.001761638230154, 0.034716203531329086, 6835.459587089728], [0.98284116664767, 0.025857190319630748, 6513.610474444275, 0.9708736757927514, 0.03056353600893742, 6458.215976918462], [1.0161553095210387, 0.025836525197028042, 7022.358947889385, 1.0217815242813324, 0.018468026494628455, 7014.982725133799]], 'E': 0.3221498813310646, 'sig_E': 0.012196246684458, 'cs': 2.2398382821025774e-15, 'sig_cs': 1.5157202180883223e-15, 'Rsq': 0.9997197958871469, 'tau at input T': 0.0068921932253929305, 'sig_tau at input T': 0.007149922746399282}, ((68, 67), 'CENel2', 0): {'T': [210.0], 'tau': [0.00025950328074446935], 'sigma_tau': [6.817664846393487e-06], 'cap': [[0.9793340852978506, 0.01853336317337582, 12243.401147887931, 0.9758375262393606, 0.013695770128982364, 12152.746634353764]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((98, 33), 'LHSel3', 0): {'T': [180.0, 190.0, 210.0, 220.0], 'tau': [8.585979599114614e-05, 2.98817527403385e-05, 4.730882652422409e-06, 2.133122453083575e-06], 'sigma_tau': [1.7355167513963759e-06, 9.187339229181331e-07, 1.2803361991241665e-07, 6.922254519708556e-08], 'cap': [[0.9701661703310397, 0.020847002000957637, 12431.021001270477, 0.9650175644609517, 0.004695463044529454, 22900.556993396654], [0.9889487179353089, 0.012166174946216215, 12889.981239166518, 1.0064284078336874, 0.007359815927427905, 22950.58889746728], [0.9915947555762583, 0.024424416303744093, 12529.666172004407, 0.9871555125065361, 0.010049878555214215, 24088.628678639052], [1.0284353242822677, 0.013476081553712842, 12609.480903822096, 1.031763827330841, 0.012705183925680385, 25807.311833923002]], 'E': 0.28077821309253576, 'sig_E': 0.002766798615593201, 'cs': 1.2469607632106575e-14, 'sig_cs': 2.068936107409238e-15, 'Rsq': 0.9999878295673044, 'tau at input T': 8.597280966145418e-05, 'sig_tau at input T': 2.0943954106837176e-05}, ((98, 33), 'LHSel3', 1): {'T': [210.0], 'tau': [0.00019990144050614507], 'sigma_tau': [5.857543720228127e-05], 'cap': [[0.5852625259827944, 0.025413571342843517, 12529.666172004407, 0.9792645182676046, 0.01069322362536408, 24088.628678639052]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((41, 15), 'CENel3', 0): {'T': [200.0, 210.0, 220.0], 'tau': [0.0007009979399416166, 0.000250841052888874, 0.00010419399187062736], 'sigma_tau': [3.476567115367395e-05, 1.2536896661412712e-05, 5.211836739426717e-06], 'cap': [[1.0227972770229765, 0.046342074665518776, 23158.396360628598, 0.9916627995076703, 0.012854031190999952, 6541.28733668556], [0.968427146361473, 0.03094472975666032, 23058.41340611829, 0.9811441124315372, 0.029393209010051723, 6770.865110931485], [1.028956599020147, 0.03459101613447377, 24979.454910396194, 1.0271020556566783, 0.03213247058363874, 7161.352231839326]], 'E': 0.3250218172961829, 'sig_E': 0.013491061362156161, 'cs': 2.66169968274676e-15, 'sig_cs': 1.9893669403807078e-15, 'Rsq': 0.9991807460846615, 'tau at input T': 0.006979527666845146, 'sig_tau at input T': 0.00800398963915761}, ((77, 90), 'CENel4', 0): {'T': [190.0, 200.0, 210.0, 220.0], 'tau': [0.0020814890512671556, 0.0006949384564573144, 3.6937888460655295e-06, 5.2083441923187425e-05], 'sigma_tau': [3.861243291809584e-05, 1.855322800823245e-05, 1.1154650428516728e-06, 5.092302886525906e-05], 'cap': [[1.0068657227357243, 0.013846004621670007, 12468.365046102475, 0.9933797131228466, 0.016979254682418097, 13101.254717121625], [1.0097026770752195, 0.017946644975327174, 12446.773103921285, 1.010719688206765, 0.026320075498039955, 13158.06667430302], [0.5815052904784785, 0.00750203501789568, 12223.666033022497, 0.982467051462249, 0.011819882477937091, 12646.25133562455], [0.6069308410650982, 0.018721312186112882, 12500.0, 1.0318633728644655, 0.011695078041634626, 12838.543736827349]], 'E': 0.33613698919617574, 'sig_E': 1.3216837905320402, 'cs': 4.999999999963583e-14, 'sig_cs': 3.679700867439873e-12, 'Rsq': -1.060981022737784, 'tau at input T': 0.000760724491930517, 'sig_tau at input T': 0.08565027702761034}, ((50, 50), 'LHSel2', 0): {'T': [190.0], 'tau': [0.0027829706420741456], 'sigma_tau': [0.0007829390766742128], 'cap': [[1.7405907261362836, 0.01235193925754802, 12811.349315770118, 1.0070518733615752, 0.012694792663854045, 12613.123673361803, 1.003837436167412, 0.009008834792604477, 22515.838594769433]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((98, 33), 'RHSel2', 0): {'T': [190.0, 200.0, 220.0], 'tau': [0.002022053462349325, 0.0006964871546015045, 0.00010480351622016112], 'sigma_tau': [4.8328413234377966e-05, 2.0948168174806346e-05, 3.3620479222544073e-06], 'cap': [[0.9879097364318926, 0.007395615330233522, 12263.02078779486, 1.0026360274602382, 0.02064812242680001, 12889.981239166518, 0.9971551896242815, 0.007544175440029376, 22950.58889746728], [1.0335318614736364, 0.01843945398631407, 12925.071054206808, 0.9843368880546158, 0.024881249956034447, 13479.913785461442, 1.0179216144340832, 0.01042105033808402, 23973.881030965356], [1.004245377398018, 0.016400564984216263, 12374.202624379986, 1.0197000518113306, 0.022970083490503505, 12609.480903822096, 1.0259794197951309, 0.011789617456983693, 25807.311833923002]], 'E': 0.31971761162927465, 'sig_E': 0.0047831421788794804, 'cs': 1.969450597569089e-15, 'sig_cs': 5.4805205452299615e-16, 'Rsq': 0.999878464221706, 'tau at input T': 0.00670082381506076, 'sig_tau at input T': 0.0027832936110877104}, ((89, 2), 'LHSel4', 0): {'T': [190.0], 'tau': [0.0017967369836589578], 'sigma_tau': [0.0017684268283483023], 'cap': [[0.9833577826878003, 0.016378535558277625, 12114.280496729163, 1.7329319156296237, 0.01831376700226948, 12798.34463825076, 1.0024042712940975, 0.01026729427025589, 22921.35615886913]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((89, 2), 'RHSel3', 0): {'T': [190.0, 200.0, 220.0], 'tau': [2.3254867392794558e-05, 0.0006864791244557248, 0.00010697088288379983], 'sigma_tau': [6.425518097600981e-06, 1.5543277712707672e-05, 3.8427678701604225e-06], 'cap': [[0.5835498010560131, 0.010486637810519265, 12798.34463825076, 0.9930558120474726, 0.010018614670760102, 22921.35615886913], [1.0292885717413427, 0.010731656440088623, 12844.30727995701, 1.0012111625330045, 0.020811882952262845, 23712.403261823187], [1.0056315287220787, 0.019464016330096333, 12147.982191436822, 1.0357774385533571, 0.009297544141418524, 25450.14402860468]], 'E': 4.366798154466189e-14, 'sig_E': 0.3029803196464168, 'cs': 1.009898841303482e-22, 'sig_cs': 1.9756667060277807e-21, 'Rsq': -1.5667165882417544, 'tau at input T': 0.00014603660470100666, 'sig_tau at input T': 0.0040371997173409705}, ((77, 90), 'LHSel1', 0): {'T': [190.0], 'tau': [2.2527309405107393e-05], 'sigma_tau': [6.651376640120325e-06], 'cap': [[0.5922310892124644, 0.013432576137423878, 12468.365046102475, 0.9971509625748092, 0.0046874150396498445, 21217.9481639698]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((60, 80), 'RHSel1', 0): {'T': [190.0, 200.0], 'tau': [0.0015429101283332216, 0.000545424105881672], 'sigma_tau': [0.0004494553795020039, 0.0001453003610176715], 'cap': [[0.5724909893066951, 0.013178832369984989, 6339.2181844754, 1.0099668381129365, 0.005459397580008749, 22540.05129886402], [0.5903359469979521, 0.0072663239596604205, 6574.8603338000075, 1.0294051630922094, 0.010216763552400117, 22688.317477978337]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((10, 10), 'RHSel3', 0): {'T': [180.0, 220.0], 'tau': [6.73610538185689e-05, 8.276067937958326e-05], 'sigma_tau': [1.9036392811290696e-05, 2.374880143907891e-05], 'cap': [[0.5574141506523225, 0.011638347916704761, 6570.369420949126, 0.9714084750940062, 0.006492260467002517, 22580.26088282659], [0.6029128252403795, 0.010568165771708568, 7801.101816846589, 1.025766376157637, 0.009448988134519224, 25699.651587944023]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((10, 10), 'RHSel3', 1): {'T': [180.0, 220.0], 'tau': [0.005373533520275411, 1.7137067361543477e-06], 'sigma_tau': [0.0014646869354770897, 4.443344788371545e-07], 'cap': [[0.5408878269473285, 0.011162720424411132, 6570.369420949126, 0.9634752559257151, 0.00626758048633326, 22580.26088282659], [0.5885918086495865, 0.012011710741235675, 7801.101816846589, 1.0270371922175168, 0.010266670242161284, 25699.651587944023]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((89, 2), 'RHSel3', 1): {'T': [220.0], 'tau': [1.690747598425948e-06], 'sigma_tau': [4.853075462159266e-07], 'cap': [[0.580069369977539, 0.021821432986309536, 12147.982191436822, 1.0153059602132006, 0.01007006588513258, 25450.14402860468]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((10, 10), 'LHSel4', 1): {'T': [200.0], 'tau': [0.0007050133804254573], 'sigma_tau': [3.3089991540609744e-05], 'cap': [[1.0221152028310423, 0.01552948260893628, 13456.245447003697, 1.0203550992946775, 0.04958831096459034, 7083.715422301953, 1.0248241061432104, 0.008829128809862561, 24067.765267526112]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}, ((50, 50), 'RHSel1', 0): {'T': [200.0], 'tau': [0.000705041548754174], 'sigma_tau': [2.7208886504268568e-05], 'cap': [[1.0050097848650459, 0.0185093518290603, 12500.0, 1.026313420298651, 0.03421406819371357, 23074.731038401966]], 'E': None, 'sig_E': None, 'cs': None, 'sig_cs': None, 'Rsq': None, 'tau at input T': None, 'sig_tau at input T': None}}


None nan
None nan
None nan
None nan
None nan
None nan
0.3221498813310646 0.3221498813310646
2.2398382821025774e-15 2.2398382821025774e-15
0.0068921932253929305 0.0068921932253929305
None nan
None nan
None nan
0.28077821309253576 0.28077821309253576
1.2469607632106575e-14 1.2469607632106575e-14
8.597280966145418e-05 8.597280966145418e-05
None nan
None nan
None nan
0.3250218172961829 0.3250218172961829
2.66169968274676e-15 2.66169968274676e-15
0.006979527666845146 0.006979527666845146
0.33613698919617574 0.33613698919617574
4.999999999963583e-14 4.999999999963583e-14
0.000760724491930517 0.000760724491930517
None nan
None nan
None nan
0.31971761162927465 0.31971761162927465
1.969450597569089e-15 1.969450597569089e-15
0.00670082381506076 0.00670082381506076
None nan
None nan
None nan
4.366798154466189e-14 4.366798154466189e-14
1.009898841303482e-22 1.009898841303482e-22
0.00014603660470100666 0.00014603660470100666
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan
None nan

Reply:
@kjl0025 I don't think I'm going to have time to track this bug down and I know that you can't reproduce it. I know that my drp output isn't 100% exactly the same. Some things are off by 1e-8, some 1e-9, some 6e-5. Given the upcoming deadline, I recommend we go with my suggestion above for now (we're still in agreement to the 0.01% level) and make an issue about it.

Reply:
Sure, that sounds good to me!  I'll make an issue for it.